packages/web-frontend/src/selectors/measureSelectors.js

```diff
@@ -34,139 +32,143 @@ const displayInfoCache = createCachedSelector(
   measureData.find(val => val.organisationUnitCode === code);
 
 const selectDisplayInfo = (measureOptions, hiddenMeasures, measureData, organisationUnitCode) =>
   safeGet(displayInfoCache, [measureOptions, hiddenMeasures, measureData, organisationUnitCode]);
 
-export const selectCurrentMeasureIds = createSelector(
-  [selectCurrentMapOverlayIds, state => state.mapOverlayBar.mapOverlayHierarchy],
-  (currentMapOverlayIds, mapOverlayHierarchy) => {
-    const hierarchy = flattenMapOverlayHierarchy(mapOverlayHierarchy);
-
-    return hierarchy.map(mapOverlay =>
-      currentMapOverlayIds.includes(mapOverlay.mapOverlayId) ? mapOverlay.measureIds : [],
-    );
+export const selectMeasureOptions = createSelector(
+  [state => state.map.measureInfo, (_, mapOverlayIds) => mapOverlayIds],
+  (measureInfo = {}, mapOverlayIds) => {
+    const selectedMeasureOptions = mapOverlayIds.reduce((results, mapOverlayId) => {
+      const { measureOptions = [] } = measureInfo[mapOverlayId] || {};
+      return [...results, ...measureOptions];
+    }, []);
+
+    const measureOptionsByKey = selectedMeasureOptions.reduce((results, measureOption) => {
+      const { key } = measureOption;
+      return results[key] ? results : { ...results, [key]: measureOption };
+    }, {});
+
+    const filteredMeasureOptions = Object.values(measureOptionsByKey);
+    return filteredMeasureOptions.length > 0 ? filteredMeasureOptions : undefined;
   },
 );
 
-export const selectDisplayedMeasureIds = createSelector(
-  [state => state.map.displayedMapOverlays, state => state.mapOverlayBar.mapOverlayHierarchy],
-  (displayedMapOverlays, mapOverlayHierarchy) => {
-    const hierarchy = flattenMapOverlayHierarchy(mapOverlayHierarchy);
-
-    return hierarchy
-      .map(mapOverlay =>
-        displayedMapOverlays.includes(mapOverlay.mapOverlayId) ? mapOverlay.measureIds : [],
-      )
-      .flat();
-  },
-);
-
-export const selectMeasureOptionsByDisplayedMapOverlays = createSelector(
-  [state => state.map.measureInfo.measureOptions, selectDisplayedMeasureIds],
-  (measureOptions, displayedMeasureIds) => {
-    if (!measureOptions) {
+const selectMeasureData = createSelector(
+  [state => state.map.measureInfo, (_, mapOverlayIds) => mapOverlayIds],
+  (measureInfo, mapOverlayIds) => {
+    if (!measureInfo || mapOverlayIds.length === 0) {
       return undefined;
     }
-    const displayedMeasureOptions = measureOptions.filter(({ key }) =>
-      displayedMeasureIds.includes(key),
-    );
 
-    return displayedMeasureOptions.length > 0 ? displayedMeasureOptions : undefined;
+    const measureData = {};
+    mapOverlayIds.forEach(mapOverlayId => {
+      const selectedMeasureData = measureInfo[mapOverlayId]?.measureData;
+      if (selectedMeasureData) {
+        selectedMeasureData.forEach(measure => {
+          measureData[measure.organisationUnitCode] = {
+            ...measureData[measure.organisationUnitCode],
+            ...measure,
+          };
```

Comment:
`...` creates a new reference/value for each spread field, which can slow down things significantly when done a few thousand times (I've seen this happening in the past).

A way to avoid this could be something like the following:
```suggestion
          measureData[measure.organisationUnitCode] = Object.assign(
            measureData[measure.organisationUnitCode] || {},
            measure,
          );
```

`...` is nice in most cases, where
* we care about immutability
* we don't care about performance

In this specific case
* we do care about performance, since each overlay can have lots of data points
* we don't care as much about immutability in `measureData`: we "own" it (it was created inside the function), and the local scope of the variable is fairly limited

Reply:
Based on the above, it would be good to confirm that `...` doesn't degrade performance here. To do so, we can select 2 overlays which both have lots of data (eg in Laos Schools) and take some measurements with `...` and with `Object.assign`.

 I'm happy for this to stay as is if there is no noticeable difference from a user perspective. `...` is commonly in modern JavaScript, and it reads nicely!

Reply:
Nice call Kostas! I've compared the difference and it seems they both rendered fast. But I would love to use Object.assign in case `...` will cost more time rendering in the future.